embassy-imxrt/src/gpio.rs

```diff
@@ -0,0 +1,1058 @@
+        drop(self);
+        Flex::<SenseEnabled>::new(new_pin)
+    }
+}
+
+/// Input pin
+pub struct Input<'d> {
+    // When Input is dropped, Flex's drop() will make sure the pin is reset to its default state.
+    pin: Flex<'d, SenseEnabled>,
+}
+
+impl<'d> Input<'d> {
+    /// New input pin
+    pub fn new(pin: Peri<'d, impl GpioPin>, pull: Pull, inverter: Inverter) -> Self {
+        let mut pin = Flex::<SenseEnabled>::new(pin);
+        pin.set_as_input(pull, inverter);
+        Self { pin }
+    }
+
+    /// Is high?
+    #[must_use]
+    pub fn is_high(&self) -> bool {
+        self.pin.is_high()
+    }
+
+    /// Is low?
+    #[must_use]
+    pub fn is_low(&self) -> bool {
+        self.pin.is_low()
+    }
+
+    /// Input level
+    #[must_use]
+    pub fn get_level(&self) -> Level {
+        self.pin.get_level()
+    }
+
+    /// Wait until the pin is high. If it is already high, return immediately.
+    #[inline]
+    pub async fn wait_for_high(&mut self) {
+        self.pin.wait_for_high().await;
+    }
+
+    /// Wait until the pin is low. If it is already low, return immediately.
+    #[inline]
+    pub async fn wait_for_low(&mut self) {
+        self.pin.wait_for_low().await;
+    }
+
+    /// Wait for the pin to undergo a transition from low to high.
+    #[inline]
+    pub async fn wait_for_rising_edge(&mut self) {
+        self.pin.wait_for_rising_edge().await;
+    }
+
+    /// Wait for the pin to undergo a transition from high to low.
+    #[inline]
+    pub async fn wait_for_falling_edge(&mut self) {
+        self.pin.wait_for_falling_edge().await;
+    }
+
+    /// Wait for the pin to undergo any transition, i.e low to high OR high to low.
+    #[inline]
+    pub async fn wait_for_any_edge(&mut self) {
+        self.pin.wait_for_any_edge().await;
+    }
+}
+
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+struct InputFuture<'d> {
+    pin: Peri<'d, AnyPin>,
+}
+
+impl<'d> InputFuture<'d> {
+    fn new(pin: Peri<'d, impl GpioPin>, int_type: InterruptType, level: Level) -> Self {
+        // Clear any existing pending interrupt on this pin
+        pin.block()
+            .intstata(pin.port())
+            .write(|w| unsafe { w.status().bits(1 << pin.pin()) });
+
+        /* Pin interrupt configuration */
+        pin.block().intedg(pin.port()).modify(|r, w| match int_type {
+            InterruptType::Edge => unsafe { w.bits(r.bits() | (1 << pin.pin())) },
+            InterruptType::Level => unsafe { w.bits(r.bits() & !(1 << pin.pin())) },
+        });
+
+        pin.block().intpol(pin.port()).modify(|r, w| match level {
+            Level::High => unsafe { w.bits(r.bits() & !(1 << pin.pin())) },
+            Level::Low => unsafe { w.bits(r.bits() | (1 << pin.pin())) },
+        });
+
+        // Enable pin interrupt on GPIO INT A
+        pin.block()
+            .intena(pin.port())
+            .modify(|r, w| unsafe { w.int_en().bits(r.int_en().bits() | (1 << pin.pin())) });
```

Comment:
I think these `modify` accesses to be guarded by a `critical_section`.

Say we have 2 instances for different pins in the same port. The first lives on the main "thread" and the other lives in an `InterruptExecutor` (interrupt "thread").

The first pin does a `read` to get the current status of the interrupt on the main thread. An interrupt comes in and execution starts on the interrupt thread. The pin on the interrupt does the RMW operation to enable it's own interrupt.

Then execution returns to the main thread. Since the modify on the first pin was interrupted, the read has the other pin's interrupt disabled. Then first pin writes to enable the interrupt which has cleared the interrupt bit of the pin on the interrupt executor.

Reply:
argh, makes sense. Fixing.

Reply:
Fixed